### ISSUE TITLE
Don't expose maven.inria.fr in Spoon parent-pom

### ISSUE THREAD

**Issue by surli** [INRIA/spoon#3216]
The Spoon snapshot repository is exposed in the parent-pom which might cause malfunction if the nexus server is down (which is the case right now) even so Spoon does not need this repository to works: it's only use for revapi comparison. It would be better to use a profile for revapi and to add the repo only in that profile.

**Comment by surli** [comment]
Or maybe we should just pushing the snapshot on maven.inria.fr an use central for that too?...

**Comment by surli** [comment]
FTR the server restarted, but looks like I informed them about it, so not something really well monitored or used it seems :)

**Comment by monperrus** [comment]
simple PR welcome :-)

**Comment by surli** [comment]
Well I'm a bit afraid to break the CI infra, I've not followed well the work done for continuous deployment... I guess best would be to push snapshots on central if it's possible, not sure what needs to be changed though.

**Comment by monperrus** [comment]
What about pushing the snapshots to https://repository.ow2.org/nexus/ instead of maven.inria.fr?

**Comment by surli** [comment]
Could be better, but I still think we should avoid to release Spoon with a reference to an external repository, especially if it's not needed for using it.

**Comment by surli** [comment]
Note we got the same problem than last year just now: apparently maven.inria.fr is down and it's blocking XWiki build since Spoon declared the repository.

**Comment by slarse** [comment]
@surli We don't use maven.inria.fr anymore, we switched to OW2 nexus a few months back. So I'm guessing this is for an older version of Spoon? But I agree that it would be better for the snapshot repository to be optional. It's a bit of an annoyance even when working as expected, as Maven searches that repository for everything else as well.

I suppose I'm the one who's most down and dirty with the CI infrastructure at the moment, so I'll see what I can do to remove the snapshot repository from the pom file. It shouldn't be much of a problem, most of the CI builds inject the snapshot repository into other projects' pom files anyway. Perhaps only the snapshot deployer job actually needs the repository declared.

**Comment by surli** [comment]
> So I'm guessing this is for an older version of Spoon?

Indeed we're still few version behind I need to upgrade but last time I tried we got issues. 

>  I'll see what I can do to remove the snapshot repository from the pom file

Well back when I created this issue, the snapshot repo was used for revapi, if it's still the case then you can probably just declare it for a specific profile and use the profile in the CI, that's what we have at xwiki for example: https://github.com/xwiki/xwiki-commons/blob/master/pom.xml#L2846

**Comment by slarse** [comment]
> Indeed we're still few version behind I need to upgrade but last time I tried we got issues.

Any idea why we don't have any XWiki builds on Jenkins? Seems like a significant client where we'll want to be alerted of regressions.

>  the snapshot repo was used for revapi

I have no knowledge of revapi, but if it's a separate project that needs snapshots, it can just declare the snapshot repository on its own? We don't have a build for revapi.

On our side of things, all [Jenkins jobs](https://ci.inria.fr/sos/) are supposed to use the snapshot repository, but how they access it differs a bit. Some expect it to be declared in the tested project, for some projects it's injected into the pom, and there are probably more variations I'm not aware of as I haven't actually peeked inside every single build job. Creating some form of order in the builds is long overdue, and fixing this at the same time would make sense.

> if it's still the case then you can probably just declare it for a specific profile and use the profile in the CI

**Comment by surli** [comment]
> Any idea why we don't have any XWiki builds on Jenkins? 

XWiki is a huge project which takes several hours for us to build entirely, so I guess it's part of the reason :) 
Now you could probably build a subpart of XWiki that uses Spoon as an integration test, we have some dedicated processors in here: https://github.com/xwiki/xwiki-commons/tree/master/xwiki-commons-tools/xwiki-commons-tool-spoon/xwiki-commons-tool-spoon-checks/src/main/java/org/xwiki/tool/spoon that are called during the build of our modules. 
I can help identify one or two modules that are using those processors and which might be interesting to run in Spoon CI if you're ok to go in that direction.

bq. I have no knowledge of revapi, but if it's a separate project that needs snapshots, it can just declare the snapshot repository on its own? We don't have a build for revapi.

So Spoon used to use revapi which is a tool to ensure the backward compatibility of the APIs of projects. Apparently it's been removed in https://github.com/INRIA/spoon/pull/3535

**Comment by slarse** [comment]
> XWiki is a huge project which takes several hours for us to build entirely, so I guess it's part of the reason :)

Ah, right. But even with a build time of several hours, it would be feasible to schedule such an undertaking something like once a week. The Jenkins nodes are mostly idle, only running each of the ~30-ish jobs once a day, most of which take only a few minutes.

> I can help identify one or two modules that are using those processors and which might be interesting to run in Spoon CI if you're ok to go in that direction.

Sure, so long as it's relatively straightforward. We don't really have the manpower to maintain anything complicated :)

> So Spoon used to use revapi which is a tool to ensure the backward compatibility of the APIs of projects. Apparently it's been removed in #3535

Ah, I see!

**Comment by monperrus** [comment]
Thanks @surli for the follow-up

> Don't expose maven.inria.fr in Spoon parent-pom

That's already done since @slarse's https://github.com/INRIA/spoon/commit/ac06d79bd8b18185d9d3e6bfd246ff1f4f071c94 

https://github.com/INRIA/spoon/pull/3860 finishes the job and closes this issue.

**Comment by monperrus** [comment]
Having the relevant part of XWiki's build process in Spoon's Jenkins is a great idea!

**Comment by slarse** [comment]
> Thanks @surli for the follow-up
> 
> > Don't expose maven.inria.fr in Spoon parent-pom
> 
> That's already done since @slarse's [ac06d79](https://github.com/INRIA/spoon/commit/ac06d79bd8b18185d9d3e6bfd246ff1f4f071c94)
> 
> #3860 finishes the job and closes this issue.

Not entirely. If I try to build Spoon from source, it will always look for all dependencies in both central and OW2 Nexus. [This is considered bad practice](https://blog.sonatype.com/2009/02/why-putting-repositories-in-your-poms-is-a-bad-idea), especially as we pull nothing from OW2. At the very least, that slows down build. A solution to this was presented by @surli in that we can just put the repository in a profile, and activate it only when necessary.

I need to do a rundown of all Jenkins jobs anyway as they are in a bit of a poor state (I found several yesterday that still try to pull snapshots from maven.inria.fr), and so now is the perfect time to make this change.

**Comment by slarse** [comment]
So, to be clear, the same problem still exists, just that it's OW2 Nexus now instead of maven.inria.fr.

**Comment by monperrus** [comment]
> A solution to this was presented by @surli in that we can just put the repository in a profile, and activate it only when necessary.

LGTM.

**Comment by slarse** [comment]
> > A solution to this was presented by @surli in that we can just put the repository in a profile, and activate it only when necessary.
> 
> LGTM.

I will get on that once I've finished with the website.

**Comment by slarse** [comment]
So, I think the best idea here is to just put the repository in a profile, and then build with breakage in the CI pipeline as it becomes apparent. I don't think I can reasonably preempt it, as I can't activate a profile that doesn't exist.

**Comment by slarse** [comment]
Now that I'm looking closer at this, I'm very uncertain why the snapshot repository is even declared in the parent pom. We pull nothing from it. To deploy to it, the entry in `<distributionManagement>` should be sufficient. It may be that a few of the Jenkins builds rely on it being declared in the parent pom, but I actually don't think so. Really, I think we can remove the snapshot repository entirely from Spoon's parent pom.

The snapshot repository is also declared in `spoon-maven-plugin`. But I changed the way the Jenkins jobs that use `spoon-maven-plugin` discover snapshots, so I don't think it's needed there either.

To summarize, unless anyone actually knows why we have the snapshot repository declared in anything but `<distributionManagement>`, I propose we remove it. And see what happens.

**Comment by slarse** [comment]
I figured it out, the extra checks for e.g. spoon-decompiler and the like, depend on the snapshot repository: https://github.com/INRIA/spoon/blob/master/chore/travis/ci-extra.sh

A profile works for handling that, I'll get on it.
